### DISCUSSION TITLE
Front end phoneme discussions

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#212] [Archived]
<USER>>>> nmstoker<br/> <DATE> [June 6, 2020, 7:11pm]<br/> <br/> Following on from the [issue](https://github.com/mozilla/TTS/issues/417)<br/> here, I thought I'd upload two files to show a few things I'm running<br/> into (and attempts to figure out a way round them)<br/> <br/> I've been using espeak-ng (*rather than espeak, given it's not<br/> maintained*) and have added a few custom words manually added to it for<br/> words that commonly appear in my training material (so as not to confuse<br/> the model training it with audio that doesn't correspond well to the<br/> phonemes provided).<br/> <br/> I mostly get good results but there are some areas where the model has<br/> not learnt well - sometimes this is down to espeak-ng quirks but several<br/> times it's the model being inconsistent. An example of the inconsistency<br/> with one of my better models is with 'make': at the start of short<br/> sentences, it's clearly said as I would say it but for longer sentences<br/> it takes on a weird 'maeek' kind of twist. That seems like something I<br/> need to fix with perhaps adding more training example audio (and<br/> verifying the quality of those samples which include make, just in case<br/> there's an issue).<br/> <br/> I started looking at how rhyming sets of words sounded to see if that<br/> distinguished flaws in the phonemes for particular words or for all<br/> words with a phoneme. I've found a few problem words that way but for<br/> the most part it seems to be that one or two particular phonemes just<br/> don't get learnt properly - I still have more investigation to pin this<br/> down (but again it seems likely I'll need to look at the training<br/> audio). An example in the output audio file attached is with the line<br/> near the beginning for '**push, bush, cush, tush**.' which generate<br/> these phonemes from espeak-ng: pˈʊʃ,bˈʊʃ,kˈʊʃ,tˈʊʃ and yet they *all*<br/> sound malformed.<br/> <br/> One area I did see espeak-ng have problems was with a whole class of<br/> long 'a' sounds. I'm using it with the UK-RP accent setting and it does<br/> make an adjustment (eg cast - *slash* > c-AH-st), but it overfitted this and you<br/> end up with pl-AH-stic, dr-AH-stic etc (which no-one says except for<br/> comedy value!) There probably is a way to fix this collectively in<br/> espeak-ng but it seemed fiddly so I ended up simply overriding each of<br/> the words with issues.<br/> <br/> Lastly I've included some initial work with heteronyms I've been playing<br/> around with. Espeak-ng handles these to an extent (eg adjusting read to<br/> the correct tense in some cases) but it wasn't quite consistent enough<br/> for my liking, so I looked into approaches to handle them separately. My<br/> rather simplistic attempt is to use a language model (KenLM) where I've<br/> manually marked the distinct word uses (appending ' *slash* |1', ' *slash* |2' etc for<br/> the distinct forms). Whilst some of the heteronyms could have been<br/> handled by POS analysis I found this wasn't that reliable and there are<br/> enough where the POS can't distinguish alone (ie where both variants are<br/> say a noun) so the language model seems better able to help in such<br/> scenarios. As I say, it's not that sophisticated so I expect smarter<br/> methods could easily out-perform it.<br/> <br/> The way it works for a heteronym where there are two variants is to<br/> generate two versions of a sentence (with the heteronym word in the<br/> sentence tagged with the relevant *slash* |1 or *slash* |2 in each corresponding<br/> sentence) and then the LM simply gives the sentence probability, with<br/> the more probable one being used to overwrite the phonemes that<br/> espeak-ng gives. The way that overwriting works *could* work for a more<br/> general correction method for bad phonemes, but the issue I see with it<br/> is the fiddly nature of deploying this (I don't mind messing with a<br/> compiles espeak-ng dictionary but it's not viable if people are going to<br/> install this smoothly with pip!) *slash* <br/> [audio1.zip](https://discourse.mozilla.org/uploads/short-url/cR4rbmQstCwrfqqmaGXYvc1dzKr.zip)<br/> (3.7 MB)<br/> [Initial_demo_of_heteronym_examples_06Jun2020.zip](https://discourse.mozilla.org/uploads/short-url/1fomXKaZCaKrAsVAglX7BnHCSDk.zip)<br/> (3.1 MB)<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/front-end-phoneme-discussions]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> repodiac<br/> <DATE> [June 8, 2020, 6:54am]<br/> <br/> , interesting research<br/> you have been doing here - fortunately, German does not have such<br/> 'heteronyms' afaik<br/> But I find your LM way of doing very intriguing: So, how do you exactly<br/> mark your sentences with ' *slash* |1' or ' *slash* |2' then, manually in advance? And<br/> so KenLM is another preprocessing step in training/inference where it<br/> replaces the exact phoneme representation for words with ' *slash* |1' etc.<br/> based on the most likely version according to its LM after espeak-ng has<br/> processed the sentence?<br/> <br/> According to your extensive tutorial on espeak-ng handling here in the<br/> discourse forum, I have assembled a special espeak-ng dictionary for<br/> German lean words (mostly from French or English), around 10k words from<br/> wiktionary, where the phoneme pattern is overwritten, simply. This works<br/> quite well. And training - as you mentioned also - is faster and more<br/> stable! It also executes quite quickly, so except for highly<br/> time-constrained web services or the like I think it's a great solution<br/> for TTS preprocessing. Thanks again for sharing your insights!<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [June 8, 2020, 9:22am]<br/> <br/> Thx for sharing your work. Yes the front-end part is a bit complicated<br/> especially for languages like English. Some alternative solutions might<br/> be:<br/> <br/> model can learn the positional pronunciation differences if they are<br/> available in the dataset. This is done by couple of different TTS<br/> papers before.<br/> <br/> since they work on phoneme aligned datasets you have more control<br/> over the whole system but then it mainly relies on the front-end so<br/> everything there should work perfectly. This is also probably why<br/> the most of the commercial providers use such an approach. Downside<br/> is that, these models does not sound as natural.<br/> <br/> figured before, as you replace components of the system with models<br/> you lose the control in fine details.<br/> <br/> corpus to create a TTS dataset. Then the model hurdles all these by<br/> itseld.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [June 8, 2020, 9:58pm]<br/> <br/> glad the German training's<br/> going well.<br/> <br/> Regarding the heteronyms you're right: the KenLM is a step in my<br/> preprocessing (the last one).<br/> <br/> The method is probably easiest to see in a screenshot:<br/> <br/> Taking the input sentence it creates permutations of each variant of<br/> each heteronym present and runs that through KenLM. So in the case above<br/> the word 'does' is found three times in the sentence, so each of those<br/> is tried with the three heteronyms for 'does' (most commonly the verb,<br/> like 'he does', but also as in 'hair does'/'fancy does' and lastly as in<br/> deer) and that creates 27 permuations to test in the KenLM model and it<br/> reckons here that the one for variant 8 (1, 3, 2) is most likely (-32.2<br/> is highest score) so it picks that and then replaces the first instance<br/> with the phonemes for 1 (ˈdʌz), the second with those for 3 (ˈduːz) and<br/> the third with those for 2 (ˈdoʊz). It's not perfect but works<br/> reasonably - you generally want a decent amount of data for each<br/> heteronym word and it has to cover a range of different word styles<br/> without letting one variant of the heteronym get too massively<br/> outnumbered.<br/> <br/> - not sure if you saw but<br/> there's a DeepMind paper out today ([blog<br/> post](https://deepmind.com/research/publications/End-to-End-Adversarial-Text-to-Speech)<br/> / [Arxiv](https://arxiv.org/abs/2006.03575)) and rather helpfully they<br/> give a bit of detail on their phoneme handling in the Appendix (Appendix<br/> E and Table 3):<br/> <br/> > Appendix E Text preprocessing<br/> ><br/> > We use phonemizer *slash* [7 *slash* ] (version 2.2) to perform partial normalisation<br/> > and phonemisation of the input text (for all our results except for<br/> > the No Phonemes ablation, where we use character sequences as input<br/> > directly). We used the espeak backend (with espeak-ng version 1.50),<br/> > which produces phoneme sequences using the International Phonetic<br/> > Alphabet (IPA). We enabled the following options that phonemizer<br/> > provides:<br/> ><br/> > • with_stress, which includes primary and secondary stress marks in<br/> > the output; *slash* <br/> > • strip, which removes spurious whitespace; *slash* <br/> > • preserve_punctuation, which ensures that punctuation is left<br/> > unchanged. This is important because punctuation can meaningfully<br/> > affect prosody.<br/> ><br/> > The phoneme sequences produced by phonemizer contain some rare symbols<br/> > (usually in non-English words), which we replace with more frequent<br/> > symbols. The substitutions we perform are listed in Table 3. This<br/> > results in a set of 51 distinct symbols.<br/> <br/> I realise that just because *they* chose that path doesn't mean it's the<br/> best<br/> but it's interesting to see that with all their resources they stuck<br/> with espeak-ng & phonemizer. Maybe for some languages it'll be less of<br/> an issue but I do think English is pretty awkward and the rules for edge<br/> cases can be hard to infer from text alone.<br/> <br/> I'll give it some more thought but in the short-term I plan to look a<br/> bit more at the phoneme output for my training data as I'd started<br/> already and see how my audio samples are for the less strong cases; I<br/> may have a go with the kind of mapping of the rare symbols they mention<br/> too to see what benefits I can gain.<br/> <br/> [](https://discourse.mozilla.org/t/gruut-discussions/71087)<br/> <br/> [Gruut discussions<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> repodiac<br/> <DATE> [June 9, 2020, 7:33am]<br/> <br/> > that creates 27 permuations to test in the KenLM model and it reckons<br/> > here that the one for variant 8 (1, 3, 2) is most likely (-32.2 is<br/> > highest score) so it picks that and then replaces the first instance<br/> > with the phonemes for 1 (ˈdʌz), the second with those for 3 (ˈduːz)<br/> > and the third with those for 2 (ˈdoʊz).<br/> <br/> Wow - fully automatic! Impressive. I am not familiar with KenLM, but I<br/> still don't understand (probably the source code would help but I don't<br/> want to bother you) how KenLM knows what your suffix marks (e.g. *slash* |1)<br/> mean? And how do you come up with the pattern, which words have to be<br/> marked up?<br/> <br/> When you have filtered the most probable setting/permutation (according<br/> to KenLM's score), where do you get the CORRECT phoneme pattern for the<br/> heteronym from and how do you find the precise position in the espeak-ng<br/> phoneme translation for replacing the original (i.e. maybe wrong)<br/> phoneme with the phoneme from KenLM's highest scored permutation?<br/> <br/> But really impressive - btw. do you do this as a professional or is it<br/> your freetime activity?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> repodiac<br/> <DATE> [June 9, 2020, 7:35am]<br/> <br/> > I realise that just because *they* chose that path doesn't mean it's<br/> > the best<br/> > but it's interesting to see that with all their resources they stuck<br/> > with espeak-ng & phonemizer.<br/> <br/> Oh, and really thanks for the pointer! Sounds good<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [June 9, 2020, 9:46am]<br/> <br/> - it's done with some<br/> Python, it's not part of KenLM, I merely call KenLM to get the sentence<br/> score. I've got a big list of heteronyms where I gathered a definition<br/> and phonemes manually for each variant, so that's used to display the<br/> definition during 'user testing' and to supply the phonemes that get<br/> swapped in. The code is a bit of a mess but it works okay.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> repodiac<br/> <DATE> [June 9, 2020, 6:14pm]<br/> <br/> Okay, but how does KenLM know that you specify different meanings of a<br/> heteronym, for instance: How does it know that does *slash* |1 is different from<br/> does *slash* |2? I also would expect that these suffixes are not part of the<br/> input to KenLM?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [June 9, 2020, 7:59pm]<br/> <br/> I have fed KenLM a file I produced where the words do indeed have *slash* |1 or<br/> *slash* |2 etc tagged on the end. Thus it will start to view sentences with<br/> similar context to those for *slash* |1 when the word is tagged *slash* |1 as more<br/> probable than if that same context was present but the word was tagged<br/> *slash* |2.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> repodiac<br/> <DATE> [June 10, 2020, 8:15am]<br/> <br/> I see. So this requires still a significant amount of (manual) labour to<br/> come up with the right tags for each heteronym AND a 'right'<br/> distribution of examples of those for training the KenLM language model,<br/> I suppose!?<br/> <br/> [Archived Post]<br/>
